From 08e12b58f53c62f6b027b5f86764281dc711a9e2 Mon Sep 17 00:00:00 2001
From: Prarit Bhargava <prarit@redhat.com>
Date: Thu, 28 Aug 2014 14:23:41 -0400
Subject: [cpufreq] Do not allow ->setpolicy drivers to provide ->target

Message-id: <1409235879-1283-128-git-send-email-prarit@redhat.com>
Patchwork-id: 89531
O-Subject: [RHEL7.1 PATCH BZ 1134639 127/185] cpufreq: Do not allow ->setpolicy drivers to provide ->target
Bugzilla: 1134369
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Mark Langsdorf <mlangsdo@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1134639

commit 9832235f3f880ab2e58141885e3b5ed3b158c0fb
Author: Rafael J. Wysocki <rafael.j.wysocki@intel.com>
Date:   Wed Mar 19 12:48:30 2014 +0100

    cpufreq: Do not allow ->setpolicy drivers to provide ->target

    cpufreq drivers that provide the ->setpolicy() callback are supposed
    to have integrated governors, so they don't need to set ->target()
    or ->target_index() and may confuse the core if any of these callbacks
    is present.

    For this reason, add a check preventing ->setpolicy cpufreq drivers
    from registering if they have non-NULL ->target or ->target_index.

    Signed-off-by: Rafael J. Wysocki <rafael.j.wysocki@intel.com>
    Acked-by: Viresh Kumar <viresh.kumar@linaro.org>

Cc: Lenny Szubowicz <lszubowi@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/cpufreq/cpufreq.c b/drivers/cpufreq/cpufreq.c
index 87d61ec..d955397 100644
--- a/drivers/cpufreq/cpufreq.c
+++ b/drivers/cpufreq/cpufreq.c
@@ -2296,7 +2296,9 @@ int cpufreq_register_driver(struct cpufreq_driver *driver_data)
 
  if (!driver_data || !driver_data->verify || !driver_data->init ||
      !(driver_data->setpolicy || driver_data->target_index ||
-      driver_data->target))
+      driver_data->target) ||
+      (driver_data->setpolicy && (driver_data->target_index ||
+      driver_data->target)))
   return -EINVAL;
 
  pr_debug("trying to register driver %s\n", driver_data->name);
-- 
1.7.1